### ISSUE TITLE
Public members Fl::awake_ring_*_ should be private (STR 3323, 3327)

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#559]
This is a copy of [STR 3327](https://www.fltk.org/str.php?L3327) which will be closed. Further discussions should be here.

Details, referring to previous comment numbers in [STR 3323](https://www.fltk.org/str.php?L3223):

- comment 4 (@Albrecht-S): `Fl::awake_ring_head_`, `Fl::awake_ring_tail_`, etc. should be private. These member variables are currently public.

- comment 5 (@imaclmaca): In 1.4.x we could implement my "fix" by adding a "check ring buffer" method in `Fl_lock.cxx`, rather than doing it in `fl_win32.cxx` as my current plan does...

- comment 9 (@Albrecht-S): Making these member variables private should be done in 1.4.0. We need to add one or more accessor methods (read-only getter's). The fact that a lock is used internally makes it IMHO **_necessary_** to prohibit direct access. WRT public API: if anybody really used these variables we must break that code. Really. We can still think about which getter methods we need to help.

**Comment by Albrecht-S** [comment]
Note: this is not really a bug (I classified it as "enhancement" but it has high priority and must be implemented in 1.4.0 (next ABI breaking version) because it does break the ABI.

Needs investigation...

**Comment by imaclmaca** [comment]
I think we should be able to make these private and provide an accessor (possibly even just under Win32) that does the necessary work for my old "fix" here.

Historically (IIRC around the 1.3.4 release) we hit a bug where the Win32 events were getting "queued up" during window drags and the like, and the queue wasn't subsequently flushed properly and things got out of synch. (Actually, this had probably always happened, but the effect was usually very subtle so we didn't notice until I hit a pathologically bad case...)
At the time the awake_ring was "visible" to the Win32 platform code directly so I adjusted the Win32 event handling to make sure the queue would be flushed - and TBH the mechanism is possibly over zealous now, it probably often tries to "flush" the queue, even when there is only a single event pending, which is harmless but perhaps a little bit inefficient? Not sure...

Anyway, as Albrecht says, these awake event queues are not the sort of thing that should be exposed and no user code should be accessing them directly at all - the Win32 code does need to be able to do the "flush", but we can provide some sort of accessor in the code that the Win32 handler can call that will perform that operation without publicly exposing these elements to the wider world.

Or... I suppose we could create the necessary linker scripts to hide these symbols, even if we leave them in public scope? That's *possible*, but would be fairly awkward to maintain since we'd need different mechanisms for the different compiler types...

**Comment by ManoloFLTK** [comment]
This should be fixed with commit 59be6a7.
@Albrecht-S please, close issue if you confirm.

**Comment by Albrecht-S** [comment]
@ManoloFLTK @imaclmaca @MatthiasWM

I see two issues with commit 59be6a7ef905eaf4adbe1fffd517a02ff64ae011:
1. the new public method name is ugly and unintuitive. I wrote a comment to the commit but I'd like to discuss this further here.
2. I believe that we should use the awake ring lock to access it

Regarding 1: I propose `Fl::awake_ring_empty()` rather than `Fl::equal_awake_ring_ends()`.

Regarding 2: This is the reason why I didn't commit this directly. Since this is a public method it would likely be necessary to use the awake ring lock to access head and tail, even if such a comparison is unlikely to be interrupted by another thread (but it's more likely to do the comparison while the ring is locked for modification).

Note: the code that modifies the awake ring uses `Fl::system_driver()->lock_ring()` and `Fl::system_driver()->unlock_ring()` to lock it. I believe we should do this here as well.

Matt, Ian, what do you think?

**Comment by ManoloFLTK** [comment]
OK for naming the new function Fl::awake_ring_empty().

About whether to use the awake ring lock : doing that adds to what currently exists in class Fl_WinAPI_System_Driver which doesn't lock the awake ring before making the comparison. I'd say the justification "since this is a public method" is debattable because I see no intention here to create a new entry in libfltk's public API. On the contrary, several previously accessible member variables of class FL become private. The new method is made public just because it's used by Fl_WinAPI_System_Driver::wait(double) which is not a member function of class Fl. It's kept hidden in a way by being undocumented. We could make the new method a member of class Fl_System_Driver and set it a friend of class FL to keep it completely out of public access.
On the other hand, Fl_WinAPI_System_Driver::wait(double) can _a priori_ run in the main thread and interrupt a child thread busy calling Fl::awake() and changing the awake ring. This might justify that the new Fl::awake_ring_empty() function should use the awake ring lock mechanism. I'm not aware enough of what happens with Win32 events queueing up to judge whether the locking is needed.

**Comment by Albrecht-S** [comment]
My motivation to make this (and maybe more) public getter methods was mentioned in my original post to this thread, comment 9. This was about user code using the public member variables that would now become private. I wrote "We can still think about which getter methods we need to help."

Hence I assumed that the new method would be public (not only formally but also documented etc.) to replace potential access by user code to the previously public awake ring member variables. The reason was that user code that used these variables would need suitable methods. At that point I didn't notice that you added documentation but removed it from the user docs (is this correct?).

If this new method would only be intended to be used internally (as you wrote above: "to keep it completely out of public access") then locking would still be necessary because otherwise the result would not be reliable because a user thread could call Fl::awake() [which uses the lock mechanism].

I believe we have two distinct and independent points to discuss:
1. Should the new method be in the public API (and documented) or not?
2. Is locking needed.

I'm open for any answer to point 1 but I believe the answer to point 2 is "yes".

**Comment by ManoloFLTK** [comment]
> I didn't notice that you added documentation but removed it from the user docs (is this correct?).

There might be a misunderstanding here because I never added nor removed documentation to/from that part of the code.

My intention here was to advance the FLTK 1.4 milestone. I just committed what was announced in this issue: make `Fl::awake_ring_*_` variables private. The possible need to add locking inside the Win32 wait() method was not mentionned yet. I agree locking is needed. I've now committed (df059de) code that performs locking and that puts all Fl::awake_ring_*_ variables out of the FLTK public API. I stop working on this issue and let other developers add public accessor functions to these variables if that seems useful.

**Comment by imaclmaca** [comment]
Sorry, I'm too late getting to this (had some stuff happening...)

TL;DR - I'd be fine with either version of the patch. I like the first version a bit more.

Long version:
I don't see any issue with Manolo's original patch as-is; it should be "the same" as what we had before, in effect. 

In particular, the locking issue: the "original" code had a long, rambling comment (mine) at line 494 onwards of Fl_Win32.cxx that explicitly notes that we are not locking the ring when testing the pointers, and that this is "unsafe" but then explains why it is actually OK anyway - and indeed it's been like that for years and (AFAIK) works OK.

So... we can *add* the locks (as Manolo was prompted to do for the later patch) but TBH I think that is not necessary here, and is "more" change. 
Also locking/unlocking is not free (and can be slow on Windows in my experience) so unless we need the extra safety (and I do not think that we do) then I'd be happy without the lock.

If we do add the locks, we need to update the comment in Fl_Win32.cxx too I guess.

As regards naming the method - I mostly do not care, as I hope no one outside Fl_Win32.cxx ever uses it. If I were being facetious, I might suggest calling it:  

"please_do_not_use_this_method_in_your_own_code_just_in_case_weird_stuff_happens", 

but that may be too long so let's not do that...

**Comment by ManoloFLTK** [comment]
Thanks Ian for this clarification.

I realize now I had not read properly the long "rambling comment" which explains why the locking is better not to be used. My vote is therefore, now, to remove it.

**Comment by Albrecht-S** [comment]
I had read and understood Ian's comment but my interpretation was: in **this** case where we call it internally we can get away w/o locking (it wouldn't do any harm) but in general locking would be "useful" (which I still think is correct).

When Manolo introduced the public API in his first patch my opinion was (as said above): we need locking because users calling this should get a reliable result.

With the second commit the situation is different. The new method is now out of the public API and Ian's comment applies again (in my own words): in this case we can avoid the locking because ...

[ I would have wished that we discussed this before committing. A PR would have been a better choice. ]

That said, if Ian thinks locking would be (too) expensive, particularly on Windows where this patch is made for, then I won't object removing the lock/unlock (as long as it is not in the public API) but it should be commented why we're not locking in this function.

@ManoloFLTK Regarding docs: yes, you're probably right, I wrote this off the top of my head w/o looking (again) at the changed code.

**Comment by imaclmaca** [comment]
Regards how expensive the lock is... I'll need to check, as I have not measured it "recently", and things may have changed.
The metrics I had were comparing Win7 vs Linux-(some version) vs VxWorks-6.9.
Both Linux and Win7 were much slower than VxWorks (as you'd expect) and Win7 was "significantly" slower than either of the others.
But that was then... And it may not be slow enough to matter anyway, in practice.

**Comment by imaclmaca** [comment]
OK: I did some timing tests for various mutex styles (the Win32 API provides several ways to do this...)
Testing on a (nominally) 2.2GHz i7 laptop.

The "bad" number, the one that caused me concern, was from using the object returned by a CreateMutex / WaitForSingleObject style of lock - for an uncontested lock / unlock cycle, this comes in at about 1.5us per cycle.

However, what we actually use is the Enter/LeaveCriticalSection mechanism, which is lighter weight (i.e. can not be used across processes, etc.) and for an uncontested lock / unlock cycle that times out at about 0.02us. 
Which is probably short enough not to matter, even if we have to iterate over "quite a lot" of cycles to flush the event queue out.

So it seems like leaving the lock in the code is harmless, and it seems "conceptually" more correct.

**Comment by Albrecht-S** [comment]
@imaclmaca OK, thanks for testing, Ian.

The only remaining issue would be if someone used the public members. But then we should get reports. Let's wait if this really happens.

I believe we're done and can close the issue. Any objections?

**Comment by ManoloFLTK** [comment]
> I believe we're done and can close the issue. Any objections?

No objection on my side.

**Comment by imaclmaca** [comment]
Yes, I believe it is OK

**Comment by Albrecht-S** [comment]
Thank you, Ian and Manolo.
Closing the issue now.
